### PR TITLE
Fixed Follow Routes and Error Checking

### PR DESCRIPTION
**Pull request by javidangarcia**
- Cleaned up follow and stock endpoints.
- Cleaned up error checking on frontend.
- Improved error component user interface.

https://github.com/javidangarcia/stock-zone/assets/89328228/6ea5d7e1-25d1-4d3c-889b-a14ef92ea320

### CODE REVIEW

File: backend/routes/follows.js [hunk 1]
```diff
@@ -4,56 +4,113 @@ import { Stock } from "../models/stock.js";
...
+        }
+
+        const followData = {
+            UserId: user.id,
+            StockId: stock.id
+        };
+
+        const follow = await Follow.findOne({ where: followData });
+
+        if (follow != null) {
```
Comment (vandanasub) [inline backend/routes/follows.js:65]: on line 52 you used `=== null` and on 65 used `!= null`. Would be good to be more consistent and be sure to understand the difference between `==` and `===` (or `!=` and `!==`) in JS -- [here's](https://stackoverflow.com/questions/38977829/what-is-the-difference-between-null-and-null) a quick explainer to understand the difference

**Review by vandanasub — COMMENTED** [review]
Nice job!

File: backend/routes/users.js [hunk 2]
```diff
@@ -31,41 +37,51 @@ router.post("/users/signup", async (req, res) => {
...
+                email
+            }
+        });
+    } catch (error) {
+        res.status(500).json({ error });
+    }
 });
 
 router.post("/users/logout", async (req, res) => {
+    console.log("Cookies:", req.cookies);
```
Comment (vandanasub) [inline backend/routes/users.js:82]: try to remember to not commit debug statements

File: frontend/src/components/StockData/StockData.jsx [hunk 3]
```diff
@@ -86,11 +90,15 @@ export default function StockData(props) {
                 <div className="loading">
                     <p>Loading...</p>
                 </div>
-            ) : stockNotFound ? (
+            ) : null}
+
+            {!isLoading && stockNotFound ? (
```
Comment (vandanasub) [inline frontend/src/components/StockData/StockData.jsx:95]: nice, conditional rendering logic looks good

File: backend/routes/follows.js [hunk 4]
```diff
@@ -4,56 +4,113 @@ import { Stock } from "../models/stock.js";
...
+const checkSession = (req, res, next) => {
+    const user = req.session.user;
+    if (!user) {
+        res.status(401).json({ error: "Missing Session." });
+    } else {
+        next();
+    }
+};
+
+router.use(checkSession);
```
Comment (BriptimusPrimus) [inline backend/routes/follows.js:16]: 👍 Well done!

File: backend/routes/follows.js [hunk 5]
```diff
@@ -4,56 +4,113 @@ import { Stock } from "../models/stock.js";
...
-        include: [{ model: Stock }]
-    });
+    try {
+        const follows = await Follow.findAll({
+            where: { UserId: user.id },
+            include: [{ model: Stock }]
+        });
+
+        if (follows.length === 0) {
+            return res
```
Comment (BriptimusPrimus) [inline backend/routes/follows.js:28]: Nit: This works, and it is also pretty common. However I personally think it's more clear to separate the res and the return:
```
res
  .status(404)
  .json({ error: "This user doesn't follow any stocks." });
return;
```

File: backend/routes/follows.js [hunk 6]
```diff
@@ -4,56 +4,113 @@ import { Stock } from "../models/stock.js";
 
 const router = express.Router();
 
+const checkSession = (req, res, next) => {
+    const user = req.session.user;
+    if (!user) {
```
Comment (BriptimusPrimus) [inline backend/routes/follows.js:9]: Better to use explicit checks for null and undefined than to rely on coercion of objects to booleans.

File: backend/routes/follows.js [hunk 7]
```diff
@@ -4,56 +4,113 @@ import { Stock } from "../models/stock.js";
...
+        res.status(401).json({ error: "Missing Session." });
+    } else {
```
Comment (BriptimusPrimus) [inline backend/routes/follows.js:11]: Super nit: There is nothing wrong with IF - ELSE, but usually it "looks" better to return early and avoid the use of the ELSE clause. Like you did below in the "/follows" route.

File: backend/routes/stocks.js [hunk 8]
```diff
@@ -10,32 +10,45 @@ router.post("/stocks", async (req, res) => {
...
+            const following = await Follow.findOne({
+                where: { UserId, StockId }
+            });
 
-        following
-            ? (stock.dataValues.following = true)
-            : (stock.dataValues.following = false);
+            following
+                ? (stock.dataValues.following = true)
+                : (stock.dataValues.following = false);
```
Comment (BriptimusPrimus) [inline backend/routes/stocks.js:44]: Nit: Here the stock object is being mutated. Do you know of a way of avoiding this mutation?

File: frontend/src/components/Error/Error.jsx [hunk 9]
```diff
@@ -1,24 +1,24 @@
...
+        setError(null);
+    }, [location.pathname]);
 
-        setTimeout(() => {
-            setErrorMessage("");
-            setError("");
-        }, 3000);
-
-    }, [error]);
+    const handleClose = () => {
```
Comment (BriptimusPrimus) [inline frontend/src/components/Error/Error.jsx:14]: 👍

File: frontend/src/components/Error/Error.jsx [hunk 10]
```diff
@@ -1,24 +1,24 @@
...
 
 export default function Error({ error }) {
-    const [errorMessage, setErrorMessage] = useState(error);
     const { setError } = useContext(UserContext);
+    const location = useLocation();
 
     useEffect(() => {
-        setErrorMessage(error);
+        setError(null);
+    }, [location.pathname]);
```
Comment (BriptimusPrimus) [inline frontend/src/components/Error/Error.jsx:12]: This seems a lot better than a timeout.

File: frontend/src/components/Error/Error.jsx [hunk 11]
```diff
@@ -1,24 +1,24 @@
 import "./Error.css";
-import { useState, useEffect, useContext } from "react";
+import { useEffect, useContext } from "react";
 import { UserContext } from "../App/App";
+import { useLocation } from "react-router-dom";
 
 export default function Error({ error }) {
```
Comment (BriptimusPrimus) [inline frontend/src/components/Error/Error.jsx:6]: If the variable `error` is a string, then a better name for it is "errorMessage", "errorMsg", or "message".

File: frontend/src/components/Follow/Follow.jsx [hunk 12]
```diff
@@ -21,9 +21,13 @@ export default function Follow({ ticker, stockData, setStockData }) {
                 setStockData({ ...stockData, following: !stockData.following });
             }
 
-            if (response.status === 409) {
+            if (response.status === 400 || response.status === 401 || response.status === 404 || response.status === 409) {
```
Comment (BriptimusPrimus) [inline frontend/src/components/Follow/Follow.jsx:24]: I think prettier would format this long line.

File: frontend/src/components/StocksYouFollow/StocksYouFollow.jsx [hunk 13]
```diff
@@ -10,8 +10,19 @@ export default function StocksYouFollow({ stocks, setStocks }) {
     useEffect(() => {
         const fetchStocksYouFollow = async () => {
             try {
-                const response = await axios.get("http://localhost:3000/follows", { withCredentials: true });
-                setStocks(response.data.stocks);
+                const response = await axios.get("http://localhost:3000/follows", { withCredentials: true, validateStatus: () => true });
+
+                if (response.status === 200) {
```
Comment (BriptimusPrimus) [inline frontend/src/components/StocksYouFollow/StocksYouFollow.jsx:15]: Good job addressing different use cases 👍

**Pull request merged**